onnx/defs/nn/defs.cc

```diff
@@ -3754,44 +3763,46 @@ ONNX_OPERATOR_SET_SCHEMA(
             builder.Add("present_value = Identity (PresentValue)");
           }
 
-          // If attn_mask is provided
+          builder.Add("NewKVSeqLen =  Shape <start = -2, end = -1> (PresentKey)");
+          builder.Add("AttnBiasShape = Concat <axis = -1> (QSeqLen, NewKVSeqLen)");
           float neg_inf = -std::numeric_limits<float>::infinity();
+          builder.Const1D("FloatNegInf", neg_inf);
+          builder.Const1D("ScalarZero", 0.f);
+
+          // If attn_mask is provided
           if (ctx.hasInput(3)) {
             auto* up = ctx.getInputType(3);
             if ((up == nullptr) || (!up->has_tensor_type()))
               return false;
             int64_t U = up->tensor_type().elem_type();
-            builder.Const1D("FloatInf", neg_inf);
-            builder.Const1D("ScalarZero", 0.f);
             builder.Add(
-                U == ONNX_NAMESPACE::TensorProto_DataType_BOOL ? "AttnBias = Where(attn_mask, ScalarZero, FloatInf)"
-                                                               : "AttnBias = Identity(attn_mask)");
+                U == ONNX_NAMESPACE::TensorProto_DataType_BOOL ? "AttnBiasShort = Where(attn_mask, ScalarZero, FloatNegInf)"
+                                                               : "AttnBiasShort = Identity(attn_mask)");
+            // If attn_mask has a shorter kv sequence length, we pad it to NewKVSeqLen with FloatNegInf
+            builder.Add("MaskKVSeqLen = Shape <start = -1> (attn_mask)")
+                .Add("PaddingKVSeqLen = Sub(NewKVSeqLen, MaskKVSeqLen)")
+                .Add("MaskPrefixShape = Shape <start = 0, end = -1> (attn_mask)")
```

Comment:
minor nit: may be a bit shorter to use Pad. But it's not a big deal. Not sure I understand the Cast down below, it seems redundant? Aren't we using float until the very last point, and casting to target type then?

Reply:
The reason for the CastLike is that the input mask can come in a type that is not FP32, but `MaskPadding` is created in FP32. So we need to cast it before Concat.